File: Freestylecoding.Math.FSharp.Tests/Natural/ModulusOperators.fs
namespace Natural

open Xunit
open Freestylecoding.Math

type public ModulusOperators() =
    static member Mod( left:'T when 'T :> System.Numerics.IModulusOperators<'T,'T,'T>, right:'T ) : 'T =
        'T.op_Modulus( left, right )

    [<Theory>]
    [<InlineData(  1u,  1u, 0u )>]  // Sanity
    [<InlineData(  0u,  1u, 0u )>]  // Sanity
    [<InlineData( 44u,  7u, 2u )>]  // multiple bits
    [<InlineData( 52u,  5u, 2u )>]  // rev
    [<InlineData( 52u, 10u, 2u )>]  // rev
    member public this.Sanity dividend divisor remainder =
        Assert.Equal(
            Natural( [remainder] ),
            ModulusOperators.Mod(
                Natural( [dividend] ),
                Natural( [divisor] )
            )
        )

    [<Fact>]
    member public this.Zero () =
        Assert.Equal(
            Natural( [0u] ),
            ModulusOperators.Mod(
                Natural( [20u] ),
                Natural( [10u] )
            )
        )

    [<Fact>]
    member public this.DivideByZero () =
        let exc = Record.Exception(
            fun () ->
                ModulusOperators.Mod( Natural.Unit, Natural.Zero )
                |> ignore
        )
        Assert.NotNull( exc )
        Assert.IsType<System.DivideByZeroException>( exc )

    [<Fact>]
    member public this.Big () =
        Assert.Equal(
            Natural( [0x12345678u] ),
            ModulusOperators.Mod(
                Natural( [0x75CD9046u; 0x6651AFF8u] ),
                Natural( [0x76543210u] )
            )
        )

File: Freestylecoding.Math.CSharp.Tests/Natural/ModulusOperators.cs
using System;
using System.Numerics;
using Xunit;

namespace Freestylecoding.Math.CSharp.Tests.Natural;

using static Freestylecoding.Math.CSharp.Tests.Helpers;
using Natural = Freestylecoding.Math.Natural;

public class ModulusOperators {
	private static T Mod<T>( T left, T right ) where T : IModulusOperators<T, T, T> =>
		left % right;

	[Theory]
	[InlineData( 1u, 1u, 0u )]  // Sanity
	[InlineData( 0u, 1u, 0u )]  // Sanity
	[InlineData( 44u, 7u, 2u )]  // multiple bits
	[InlineData( 52u, 5u, 2u )]  // rev
	[InlineData( 52u, 10u, 2u )]  // rev
	public void Sanity( uint dividend, uint divisor, uint remainder ) =>
		Assert.Equal(
			new Natural( remainder ),
			Mod(
				new Natural( dividend ),
				new Natural( divisor )
			)
		);

	[Fact]
	public void Zero() =>
		Assert.Equal(
			new Natural( 0u ),
			Mod(
				new Natural( 20u ),
				new Natural( 10u )
			)
		);

	[Fact]
	public void DivideByZero() {
		Exception exc = Record.Exception(
			() =>
				Mod( Natural.Unit, Natural.Zero )
		);

		Assert.NotNull( exc );
		Assert.IsType<DivideByZeroException>( exc );
	}

	[Fact]
	public void Big() =>
		Assert.Equal(
			new Natural( 0x12345678u ),
			Mod(
				FListNatural( 0x75CD9046u, 0x6651AFF8u ),
				new Natural( 0x76543210u )
			)
		);
}